scipy/stats/_bws_test.py

```diff
@@ -0,0 +1,230 @@
+import numpy as np
+from collections import namedtuple
+from ._axis_nan_policy import _axis_nan_policy_factory
+
+
+def calculate_bws_sum(ranks, nm):
+    '''Calculate the sum for the Baumgartner-Weiss-Schindler statistic'''
+    sum = 0
+    size = len(ranks)
+    factor = 1 / (size + 1)
+    for i, r in enumerate(ranks):
+        # index & ranks are 0-based, but we want 1-based
+        i1, r1 = i + 1, r + 1
+        sum += (r1 - nm / size * i1) ** 2 / (i1 * factor * (1 - i1 * factor))
+    return sum
+
+
+def bws_statistic_from_ranks(ranks, n, m):
+    '''Calculate the Baumgartner-Weiss-Schindler statistic from a set of ranks of n values from one sample.'''
+    nm = n + m
+
+    # Calculate the BWS statistic from a set of ranks of n values from one sample
+    # and m values from another sample.
+    b_x = calculate_bws_sum(ranks[:n], nm) / (m * nm)
+    b_y = calculate_bws_sum(ranks[n:], nm) / (n * nm)
+    b = (b_x + b_y) * 0.5
+
+    return b
+
+
+def shuffle(a, b):
+    '''Permutate two groups while keeping the ordering of the groups.'''
+    if not a.size:
+        yield b
+        return
+    if not b.size:
+        yield a
+        return
+    x = a[-1]
+    yield from (np.concatenate([l, [x]]) for l in shuffle(a[:-1], b))
+    x = b[-1]
+    yield from (np.concatenate([l, [x]]) for l in shuffle(a, b[:-1]))
+
+
+def bws_test_exact_from_ranks(ranks, n, m):
+    '''Exact test for the Baumgartner-Weiss-Schindler test.'''
+    b_stats = bws_statistic_from_ranks(ranks, n, m)
+
+    numlarger = 0
+    numtotal = 0
+
+    for perm in shuffle(ranks[:n], ranks[n:]):
+        perm_ranks = np.concatenate([sorted(perm[:n]), sorted(perm[n:])])
+
+        # Now we can calculate the bws again and note if it was larger than before.
+        sample_stats = bws_statistic_from_ranks(perm_ranks, n, m)
+
+        if sample_stats >= b_stats:
+            numlarger += 1
+
+        numtotal += 1
+
+    # A larger (or equal) BWS statistic value indicates samples from distributions
+    # that are "more apart" and thus that there is a difference between the samples.
+    pvalue = numlarger / numtotal
+
+    return b_stats, pvalue
+
+
+BWSResult = namedtuple('BWSResult', ('statistic', 'pvalue'))
+
+
+def bws_test_sampled_from_ranks(ranks, n, m, numsamples):
+    '''Random permutation test for the Baumgartner-Weiss-Schindler test.'''
+    b_stats = bws_statistic_from_ranks(ranks, n, m)
+
+    # Array to save b statistics
+    numlarger = 0
+
+    for _ in range(numsamples):
+        # Now shuffle ranks and then ensure they are sorted for each group.
+        np.random.shuffle(ranks)
```

Comment:
Create a RNG via `rng = np.random.default_rng()` and then use `rng.shuffle`

https://numpy.org/devdocs/reference/random/generated/numpy.random.Generator.shuffle.html#numpy.random.Generator.shuffle